erasure-readfile.go

```diff
@@ -65,128 +118,151 @@ func erasureReadFile(writer io.Writer, disks []StorageAPI, volume string, path s
+		// Current disk index from which to read, this will be used later
+		// in case one of the parallel reads fails.
 		index := 0
+
 		// Read from the disks in parallel.
 		for _, disk := range orderedDisks {
 			if disk == nil {
 				index++
 				continue
 			}
+
+			// Increment wait group.
 			wg.Add(1)
+
+			// Start reading from disk in a go-routine.
 			go func(index int, disk StorageAPI) {
 				defer wg.Done()
-				ok := bitrotVerify(index)
-				if !ok {
+
+				// Verify bit rot for this disk slice.
+				if !bitrotVerify(index) {
 					// So that we don't read from this disk for the next block.
 					orderedDisks[index] = nil
 					return
 				}
-				buf := make([]byte, curChunkSize)
-				// Note that for the offset calculation we have to use chunkSize and not
-				// curChunkSize. If we use curChunkSize for offset calculation then it
-				// can result in wrong offset for the last block.
-				n, err := disk.ReadFile(volume, path, block*chunkSize, buf)
+
+				// NOTE: That for the offset calculation we have to use chunkSize and
+				// not curChunkSize. If we use curChunkSize for offset calculation
+				// then it can result in wrong offset for the last block.
+				var buffer = new(bytes.Buffer)
+				err := copyN(buffer, disk, volume, path, block*chunkSize, curChunkSize)
 				if err != nil {
 					// So that we don't read from this disk for the next block.
 					orderedDisks[index] = nil
 					return
 				}
-				enBlocks[index] = buf[:n]
+				// Copy the read blocks.
+				enBlocks[index] = buffer.Bytes()
+
+				// Successfully read.
 			}(index, disk)
+
 			index++
-			diskCount--
-			if diskCount == 0 {
+			readableDiskCount--
+			// We have read all the readable disks.
+			if readableDiskCount == 0 {
 				break
 			}
 		}
+
+		// Wait for all the reads to finish.
 		wg.Wait()
 
-		// Count number of data and parity blocks that were read.
-		var successDataBlocksCount = 0
-		var successParityBlocksCount = 0
-		for bufidx, buf := range enBlocks {
-			if buf == nil {
-				continue
-			}
-			if bufidx < eInfo.DataBlocks {
-				successDataBlocksCount++
-				continue
-			}
-			successParityBlocksCount++
-		}
+		// FIXME: make this parallel.
 
-		if successDataBlocksCount < eInfo.DataBlocks {
-			// If we don't have DataBlocks number of data blocks we will have to read enough
-			// parity blocks such that we have DataBlocks+1 number for blocks for reedsolomon.Reconstruct()
+		// If we have all the data blocks no need to decode.
+		if !isSuccessDataBlocks(orderedDisks, eInfo.DataBlocks) {
+			// If we don't have DataBlocks number of data blocks we
+			// will have to read enough parity blocks such that we
+			// have DataBlocks+1 number for blocks for rs.Reconstruct().
 			for ; index < len(orderedDisks); index++ {
-				if (successDataBlocksCount + successParityBlocksCount) == (eInfo.DataBlocks + 1) {
-					// We have DataBlocks+1 blocks, enough for reedsolomon.Reconstruct()
+				// We have enough blocks to decode, break out.
```

Comment:
In the above loop, we intend to start reading parity blocks, given that we don't have enough data blocks to serve the request. If that is the case, then we could start the loop with ``index = dataBlocks`` instead of carrying over from previous loop, many lines above. Does that make sense?